src/main/java/org/wildfly/security/auth/realm/KeyStoreBackedSecurityRealm.java

```diff
@@ -72,7 +79,47 @@ public KeyStoreBackedSecurityRealm(final KeyStore keyStore, final Supplier<Provi
 
     @Override
     public RealmIdentity getRealmIdentity(final Principal principal) throws RealmUnavailableException {
-        return principal instanceof NamePrincipal ? new KeyStoreRealmIdentity(principal.getName()) : RealmIdentity.NON_EXISTENT;
+        if (principal instanceof NamePrincipal) {
+            return new KeyStoreRealmIdentity(principal.getName());
+        } else {
+            final X500Principal x500Principal = X500PrincipalUtil.asX500Principal(principal);
+            if (x500Principal == null) {
+                return RealmIdentity.NON_EXISTENT;
+            } else {
+                final KeyStore keyStore = this.keyStore;
+                try {
+                    final Enumeration<String> aliases = keyStore.aliases();
+                    while (aliases.hasMoreElements()) {
+                        final String alias = aliases.nextElement();
+                        if (keyStore.isCertificateEntry(alias)) {
+                            final Certificate certificate = keyStore.getCertificate(alias);
+                            if (certificate instanceof X509Certificate && X500PrincipalUtil.asX500Principal(((X509Certificate) certificate).getSubjectX500Principal()).equals(x500Principal)) {
+                                return new KeyStoreRealmIdentity(alias);
+                            }
+                        }
+                    }
+                } catch (KeyStoreException e) {
+                    throw ElytronMessages.log.failedToReadKeyStore(e);
+                }
+                return RealmIdentity.NON_EXISTENT;
+            }
+        }
+    }
+
+    @Override
+    public RealmIdentity getRealmIdentity(final Evidence evidence) throws RealmUnavailableException {
+        if (evidence instanceof X509PeerCertificateChainEvidence) {
+            // scan the keystore for a match
+            final String alias;
+            try {
+                alias = keyStore.getCertificateAlias(((X509PeerCertificateChainEvidence) evidence).getFirstCertificate());
```

Comment:
There can be no such certificate in the key store -> `new KeyStoreRealmIdentity(null)` would be returned.